### PR TITLE
ARROW-1630: [Serialization] Support Python datetime objects

### PR DESCRIPTION
**Pull request by pcmoritz**
An additional pair of eyes would be helpful, somewhat strangely the tests are passing for some datetime objects and not for others.

### CODE REVIEW

File: cpp/src/arrow/python/arrow_to_python.cc [hunk 1]
```diff
@@ -122,6 +123,12 @@ Status GetValue(PyObject* context, const Array& arr, int64_t index, int32_t type
     case Type::DOUBLE:
       *result = PyFloat_FromDouble(static_cast<const DoubleArray&>(arr).Value(index));
       return Status::OK();
+    case Type::DATE64: {
+      PyDateTime_IMPORT;
```
Comment (wesm) [inline cpp/src/arrow/python/arrow_to_python.cc]: Ideally this should be executed in the highest scope possible within this compilation unit, let me have a tinker

File: cpp/src/arrow/python/util/datetime.h [hunk 2]
```diff
@@ -69,6 +71,48 @@ static inline Status PyTime_from_int(int64_t val, const TimeUnit::type unit,
   return Status::OK();
 }
 
+static inline Status PyDateTime_from_int(int64_t val, const TimeUnit::type unit,
```
Comment (wesm) [inline cpp/src/arrow/python/util/datetime.h]: This is really gnarly, we're going to have to root around in the Python datetimemodule implementation to get this right across platforms: https://github.com/python/cpython/blob/master/Modules/_datetimemodule.c#L4326. I started looking at this but it's a can of worms. We need to also avoid UNIX y2k problems

**Comment by wesm** [comment]
Can you please rebase? I just rebased master after 0.7.1 release

**Comment by wesm** [comment]
We should try to resolve this, I can help with some of the spelunking in datetimemodule or in NumPy (which should have some of this implemented as well in a form we can adapt), let me know

**Comment by pcmoritz** [comment]
Yeah, I'll have another look tonight and let you know how it goes :)

**Comment by pcmoritz** [comment]
Looks like things are working now, one remaining problem is that PyDateTime_to_us cannot handle years before 1900 because mktime cannot deal with negative years (at least not on all platforms). Any thoughts on what it should be replaced with?

**Comment by wesm** [comment]
Hm, probably the best place to look for this would be the NumPy codebase, i.e. somewhere buried in https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/datetime.c probably or one of the related modules

**Comment by pcmoritz** [comment]
This seems to be working, thanks for the suggestion! The last remaining bit is to get rid of gmtime_r so it runs on windows; it's a little tricky to get it right with negative seconds since the epoch, will look into it later tonight.

**Comment by pcmoritz** [comment]
Timings:
```
import pyarrow
import datetime

dates = [datetime.datetime(year=1900 + i % 50, month=3, day=1, hour=i%24, second=i%60) for i in range(10000)]

%timeit x = pickle.dumps(dates)
9.8 ms ± 65.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%timeit y = pickle.loads(x)
2.31 ms ± 54.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%timeit x = pyarrow.serialize(dates).to_buffer()
14.4 ms ± 774 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%timeit pyarrow.deserialize(x)
14.2 ms ± 89.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

So performance wise there is still some room for improvement.

**Comment by pcmoritz** [comment]
This looks much better:
```
In [2]: %timeit x = pyarrow.serialize(dates).to_buffer()
3.69 ms ± 33.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [3]: x = pyarrow.serialize(dates).to_buffer()

In [4]: %timeit pyarrow.deserialize(x)
3.92 ms ± 22.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**Comment by pcmoritz** [comment]
This is now ready to merge.

cc @cpcloud

File: cpp/src/arrow/python/util/datetime.h [hunk 3]
```diff
@@ -18,22 +18,178 @@
...
 #include <datetime.h>
+#include "arrow/status.h"
+#include "arrow/util/logging.h"
 #include "arrow/python/platform.h"
 
 namespace arrow {
 namespace py {
 
+// The following code is adapted from
+// https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/datetime.c
```
Comment (wesm) [inline cpp/src/arrow/python/util/datetime.h]: Could we add the NumPy license to our LICENSE.txt at the bottom? Maybe also remove the occurrence of NumPy from the NOTICE.txt. I am not 100% sure when we have to use NOTICE and when only append to LICENSE.txt. From looking at Apache Kudu it seems most of what is in NOTICE should be in the LICENSE.txt https://github.com/apache/kudu

Comment (pcmoritz) [reply cpp/src/arrow/python/util/datetime.h]: Ok, I added it to LICENSE.txt and removed it from NOTICE.txt.

**Comment by wesm** [comment]
awesome, thanks! +1

**Pull request closed without merging**